### ISSUE TITLE
URLRequest HTTP headers Init Crash

### ISSUE THREAD

**Issue by jerometonnelier** [Alamofire/Alamofire#3425]
* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md).

## What did you do?
Just perform a regular POST request

## What did you expect to happen?

No Crash 

## What happened instead?
I often get a crash regarding URLRequest HTTP Headers.

## Alamofire Environment
**Alamofire Version:5.4.1
**Dependency Manager: SPM
**Xcode Version:12.4
**Swift Version:5
**Platform(s) Running Alamofire:iOS 14
**macOS Version Running Xcode: 11.2.3

## Demo Project

Here is the code I use to perform my requests

```
func dataRequest<T: Decodable>(_ request: RequestObject<T>) -> DataRequest {
        var headers: HTTPHeaders = HTTPHeaders()
        commonHeaders?.forEach({ headers.add($0) })
        request.headers?.forEach({ headers.add($0) })
        let dataRequest: DataRequest!
        if let data = request.multipartData {
            dataRequest = AF.upload(multipartFormData: data,
                                        to: baseURL.appendingPathComponent(request.endpoint ?? ""),
                                        method: .post,
                                        headers: headers)
        } else {
            dataRequest = AF.request(baseURL.appendingPathComponent(request.endpoint ?? ""),
                                     method: request.method,
                                     parameters: request.parameters,
                                     encoder: request.encoder,
                                     headers: headers,
                                     interceptor: nil)
        }
        printDataRequest(request: dataRequest)
        return dataRequest
    }
```

Here is the stack symbol trace

``` ▿ 40 elements
  - 0 : "0   ???                                 0x0000000119b4c5b8 0x0 + 4726244792"
  - 1 : "1   ???                                 0x0000000119b4c6ba 0x0 + 4726245050"
  - 2 : "2   taxi.Chauffeur                      0x000000010f456540 main + 0"
  - 3 : "3   taxi.Chauffeur                      0x000000010f73b5d0 $s10Foundation10URLRequestV9AlamofireE7headersAD11HTTPHeadersVvg + 144"
  - 4 : "4   taxi.Chauffeur                      0x00000001102da421 $s13KNetworkLayer3APIPAAE16printDataRequest33_DEFE19E8A11D40176BBDDD7289F7099ALL7requesty9Alamofire0eF0C_tF + 2065"
  - 5 : "5   taxi.Chauffeur                      0x00000001102d97bb $s13KNetworkLayer3APIPAAE11dataRequesty9Alamofire04DataE0CAA0E6ObjectCyqd__GSeRd__lF + 3099"
  - 6 : "6   taxi.Chauffeur                      0x00000001102de757 $s13KNetworkLayer3APIP015kPromiseNetworkB0E7performy0D3Kit0D0Cyqd__GAA13RequestObjectCyqd__GSeRd__lFyAF8ResolverCyqd__GXEfU0_ + 183"
  - 7 : "7   taxi.Chauffeur                      0x000000010f87dfc2 $s10PromiseKit0A0C8resolverACyxGyAA8ResolverCyxGKXE_tcfc + 258"
  - 8 : "8   taxi.Chauffeur                      0x000000010f87deb7 $s10PromiseKit0A0C8resolverACyxGyAA8ResolverCyxGKXE_tcfC + 55"
  - 9 : "9   taxi.Chauffeur                      0x00000001102dd977 $s13KNetworkLayer3APIP015kPromiseNetworkB0E7performy0D3Kit0D0Cyqd__GAA13RequestObjectCyqd__GSeRd__lF + 311"
  - 10 : "10  taxi.Chauffeur                      0x000000010f2e0ad4 $s14taxi_Chauffeur6AppAPIC15performAndRetry5route10PromiseKit0I0CyxG13KNetworkLayer13RequestObjectCyxG_tSeRzlF + 260"
  - 11 : "11  taxi.Chauffeur                      0x000000010f2df935 $s14taxi_Chauffeur6AppAPIC7perform5route17showMessageOnFail10PromiseKit0K0CyxG13KNetworkLayer13RequestObjectCyxG_SbtSeRzlFyAG8ResolverCyxGXEfU_ + 1093"
  - 12 : "12  taxi.Chauffeur                      0x000000010f87dfc2 $s10PromiseKit0A0C8resolverACyxGyAA8ResolverCyxGKXE_tcfc + 258"
  - 13 : "13  taxi.Chauffeur                      0x000000010f87deb7 $s10PromiseKit0A0C8resolverACyxGyAA8ResolverCyxGKXE_tcfC + 55"
  - 14 : "14  taxi.Chauffeur                      0x000000010f2df4b8 $s14taxi_Chauffeur6AppAPIC7perform5route17showMessageOnFail10PromiseKit0K0CyxG13KNetworkLayer13RequestObjectCyxG_SbtSeRzlF + 168"
  - 15 : "15  taxi.Chauffeur                      0x000000010f2e234c $s14taxi_Chauffeur6AppAPIC6addNew7vehicle10PromiseKit0H0CyAA17EmptyResponseDataCGAA7VehicleC_tF + 156"
  - 16 : "16  taxi.Chauffeur                      0x000000010f53b512 $s14taxi_Chauffeur18VehicleCoordinatorC6didAdd7vehicle10completionyAA0C0C_ySbctF + 194"
  - 17 : "17  taxi.Chauffeur                      0x000000010f53c829 $s14taxi_Chauffeur18VehicleCoordinatorCAA0cD8DelegateA2aDP6didAdd7vehicle10completionyAA0C0C_ySbctFTW + 9"
  - 18 : "18  taxi.Chauffeur                      0x000000010f2f76af $s14taxi_Chauffeur27VehicleDetailViewControllerC4saveyyF + 1087"
  - 19 : "19  taxi.Chauffeur                      0x000000010f2f7e3b $s14taxi_Chauffeur27VehicleDetailViewControllerC4saveyyFTo + 43"
  - 20 : "20  UIKitCore                           0x00007fff2469d19e -[UIApplication sendAction:to:from:forEvent:] + 83"
  - 21 : "21  UIKitCore                           0x00007fff23fc6684 -[UIControl sendAction:to:forEvent:] + 223"
  - 22 : "22  UIKitCore                           0x00007fff23fc69a7 -[UIControl _sendActionsForEvents:withEvent:] + 332"
  - 23 : "23  UIKitCore                           0x00007fff23fc5290 -[UIControl touchesEnded:withEvent:] + 500"
  - 24 : "24  UIKitCore                           0x00007fff246d984e -[UIWindow _sendTouchesForEvent:] + 1287"
  - 25 : "25  UIKitCore                           0x00007fff246db6c7 -[UIWindow sendEvent:] + 4774"
  - 26 : "26  UIKitCore                           0x00007fff246b5466 -[UIApplication sendEvent:] + 633"
  - 27 : "27  UIKitCore                           0x00007fff24745f04 __processEventQueue + 13895"
  - 28 : "28  UIKitCore                           0x00007fff2473c7e3 __eventQueueSourceCallback + 70"
  - 29 : "29  CoreFoundation                      0x00007fff2039038a __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17"
  - 30 : "30  CoreFoundation                      0x00007fff20390282 __CFRunLoopDoSource0 + 180"
  - 31 : "31  CoreFoundation                      0x00007fff2038f7c6 __CFRunLoopDoSources0 + 346"
  - 32 : "32  CoreFoundation                      0x00007fff20389f2f __CFRunLoopRun + 878"
  - 33 : "33  CoreFoundation                      0x00007fff203896d6 CFRunLoopRunSpecific + 567"
  - 34 : "34  GraphicsServices                    0x00007fff2c257db3 GSEventRunModal + 139"
  - 35 : "35  UIKitCore                           0x00007fff24696cf7 -[UIApplication _run] + 912"
  - 36 : "36  UIKitCore                           0x00007fff2469bba8 UIApplicationMain + 101"
  - 37 : "37  taxi.Chauffeur                      0x000000010f45658b main + 75"
  - 38 : "38  libdyld.dylib                       0x00007fff2025a3e9 start + 1"
  - 39 : "39  ???                                 0x0000000000000007 0x0 + 7"
```

Here is a screenshot too
<img width="2048" alt="Capture d’écran 2021-03-19 à 15 27 09" src="https://user-images.githubusercontent.com/1484522/111795566-9c795980-88c7-11eb-813b-f5ef562b7847.png">

The crash happens in the getter

```
extension URLRequest {
    /// Returns `allHTTPHeaderFields` as `HTTPHeaders`.
    public var headers: HTTPHeaders {
        get { allHTTPHeaderFields.map(HTTPHeaders.init) ?? HTTPHeaders() }
        set { allHTTPHeaderFields = newValue.dictionary }
    }
}
```

The print is a pretty regular print (it happens also if I remove the print of course)

`if let headers = request.convertible.urlRequest?.headers { print("• Headers: \(headers))") }`

**Comment by jshier** [comment]
Thanks for the report. If you run without the `print`, what does the crash look like then? In many cases, crashes that this (with no apparent cause) are caused by memory corruption and threading issue. If you run with the thread sanitizer enabled, what do you see?

**Comment by jerometonnelier** [comment]
Hi jshier ;-)
I did run it w/o the print, it crashes sometimes but it's much more difficult to reproduce. 
However it does crash once and a while, and it's problematic in production app ;-)
I'll run it with Thread Sanitizer enable and get back to you.

**Comment by jerometonnelier** [comment]
I ran it with the sanitizer enabled, same issue here :-/

**Comment by jshier** [comment]
You mean it crashed on the the `print` but without a sanitizer warning?

**Comment by jerometonnelier** [comment]
Yep

**Comment by jshier** [comment]
Ah, can you try the `master` branch real quick, this is probably fixed.

**Comment by jerometonnelier** [comment]
Ok I'll give it a try on Monday, I have to stop for now ;-) 
Thanks for your help

**Comment by jerometonnelier** [comment]
It seems fixed indeed ;-) When do you plan on releasing a news version ? Thanks for your help.

**Comment by jshier** [comment]
This fix has been released in 5.4.2.
